app/src/main/java/ch/epfl/polychef/utils/FavouritesUtils.java

```diff
@@ -0,0 +1,151 @@
+            setButton(button);
+            button.setOnCheckedChangeListener((buttonView, isChecked) -> onClickToggleButton(userStorage, button, recipe));
+        } else {
+            button.setVisibility(View.GONE);
+        }
+    }
+
+    private void onClickToggleButton(UserStorage userStorage, ToggleButton button, Recipe recipe) {
+        setButton(button);
+        setOrRemoveRecipe(userStorage, recipe, button.isChecked());
+    }
+
+    private void setButton(ToggleButton button) {
+        if(button.isChecked()) {
+            button.setBackgroundDrawable(ContextCompat.getDrawable(GlobalApplication.getAppContext(), R.drawable.ic_star_black_yellow));
+        } else {
+            button.setBackgroundDrawable(ContextCompat.getDrawable(GlobalApplication.getAppContext(), R.drawable.ic_star_black_grey));
+        }
+    }
+
+    private void setOrRemoveRecipe(UserStorage userStorage, Recipe recipe, boolean shouldAdd) {
+        List<Recipe> recipes = getOfflineFavourites();
+        if(shouldAdd) {
+            userStorage.getPolyChefUser().addFavourite(recipe.getRecipeUuid());
+        } else {
+            userStorage.getPolyChefUser().removeFavourite(recipe.getRecipeUuid());
+        }
+        userStorage.updateUserInfo();
+        if(shouldAdd) {
+            recipes.add(recipe);
+        } else {
+            recipes.remove(recipe);
+        }
+        putFavouriteList(recipes);
+    }
+
+    /**
+     * Get the locally saved favourites.
+     *
+     * @return the locally saved favourites
+     */
+    public List<Recipe> getOfflineFavourites() {
+        SharedPreferences sharedPref = getSharedPreferences();
+        String recipesJson = sharedPref.getString(favouriteKey, "");
+        return recipesJson.isEmpty() ? new ArrayList<>() : gson.fromJson(recipesJson, new TypeToken<List<Recipe>>() {
+        }.getType());
+    }
+
+    /**
+     * Add all online favourites of a user to the local storage for later use.
+     *
+     * @param user    the user
+     */
+    public void setOfflineFavourites(User user) {
+        Preconditions.checkArgument(user != null, "User can not be null");
+        List<String> favouritesList = user.getFavourites();
+        List<Recipe> newOfflineRecipes = new ArrayList<>();
+        for (int i = 0; i < favouritesList.size(); ++i) {
+            final boolean isLast = i == favouritesList.size() - 1;
+            getRecipeStorage().readRecipeFromUuid(favouritesList.get(i), new CallHandler<Recipe>() {
+                @Override
+                public void onSuccess(Recipe data) {
+                    newOfflineRecipes.add(data);
+                    if (isLast) {
+                        putFavouriteList(newOfflineRecipes);
+                    }
+                }
+
+                @Override
+                public void onFailure() {
+                    Log.e(TAG, "Error could not charge recipe");
+                }
+            });
+        }
+    }
```

Comment:
I'm not sure, but I fixed it in https://github.com/Ingloryon/SDP_PolyChef/pull/123/commits/d84378da27cf4453392cab555fec8d97770faa99 as we cannot be sure whether it will always work or not

Reply:
Yeah, better safe that sorry!